tests/basics/io_buffered_writer.py

```diff
@@ -28,3 +28,16 @@
 
 # hashing a BufferedWriter
 print(type(hash(buf)))
+
+# Test failing flush()
+class MyIO(io.IOBase):
+    def write(self, buf):
+        return None
+
+buf = io.BufferedWriter(MyIO(), 8)
+buf.write(b"foobar")
+try:
+    buf.flush()
+    print("flushed")
+except OSError:
+    print("OSError")
```

Comment:
I think this needs to test a little more, to ensure we have the logic correct for the case of a failed flush.

In particular we want to test that `self->len = 0` should not be executed if there's an error.  Because if we turn the assertion off then the change here changes semantics.

I suggest to loop over this try-except 4 times:
- the first two should fail with errors and print OSError
- the third should succeed and flush all 6 bytes (and `MyIO.write` should print those bytes out)
- the fourth should succeed and flush nothing, and not even call `MyIO.write`

To do this you can add some simple state to `MyIO` so it fails the first two calls to `write`, then succeeds the remaining calls.